### DISCUSSION TITLE
Obtaining key generation parameters on MED1775

### DISCUSSION THREAD

**Discussion by MasterCodeIT** [jglim/CaesarSuite#32] [Q&A]
Would you mind displaying "Obtaining key generation parameters" on this DLL?

I found the security access levels (5, 9, 0B) but I can not find the cryptographic constants.

![image](https://user-images.githubusercontent.com/73858519/122281383-339a8100-ceb8-11eb-8dc3-b5c1cfad81c1.png)

I can not seem to find it.

**Comment by jglim** [comment]
In brief, looking at `MED1775_MED1775_12_42_00.dll` as an example:

- From the exports, start from `GenerateKeyExOpt`. There should be paths for 3 access levels: 5, 9 and 11.
- For each case, there will usually be an alloc, parameter initialization routine, and the actual key generation routine. 
- Enter the parameter initialization routine. At this point, the security provider should be known (5,11: PowertrainSecurityAlgo, 9: DaimlerStandardSecurityAlgo)
- Look further into the child routines in parameter init; there will be a shared routine where all parameters are initialized for 5, 9 and 11. Key generation parameters are found there

As an example for level 9:

![image](https://user-images.githubusercontent.com/1116555/122319263-f59e6b00-cf52-11eb-87ef-fd4ebfc257de.png)

From that parameter, the definition can be created:

```
  {
    "EcuName": "MED1775",
    "Aliases": [],
    "AccessLevel": 9,
    "SeedLength": 8,
    "KeyLength": 4,
    "Provider": "DaimlerStandardSecurityAlgo",
    "Origin": "MED1775_MED1775_12_42_00",
    "Parameters": [
      {
        "Key": "K",
        "Value": "35A9C476",
        "DataType": "ByteArray"
      }
    ]
  }
```

---
As mentioned in jglim/CaesarSuite#29, I would appreciate if non-issues are created in its correct place as discussions [here](https://github.com/jglim/CaesarSuite/discussions) instead; I may not be able to reply to further non-issues.
